File: sdk/elixir/lib/bsdkrun/cache.ex
defmodule Bsdkrun.Cache do
  @moduledoc """
  Cached guest directories.

  Entries are keyed, so a rebuild can pick up where the last one left off:

      case Bsdkrun.Cache.restore("web", key: key, restore_keys: ["deps-"]) do
        {:ok, %{restored: false}} ->
          Bsdkrun.Sandbox.exec("web", ["npm", "ci"])
          Bsdkrun.Cache.save("web", "/app/node_modules", key: key)

        {:ok, _hit} ->
          :ok
      end

  Where entries live — host disk or S3 — is host configuration, not an SDK
  concern: set `BSDKRUN_CACHE_BACKEND` / `BSDKRUN_CACHE_S3_*`, or write
  `~/.config/bsdkrun/cache.toml`.
  """

  alias Bsdkrun.{Cli, Error}

  @typedoc "A stored cache entry, as `cache ls` reports it."
  @type entry :: %{
          key: String.t(),
          path: String.t(),
          compression: String.t(),
          size: non_neg_integer(),
          created: non_neg_integer(),
          digest: String.t()
        }

  @typedoc "What a restore did. A miss is not an error — check `:restored`."
  @type result :: %{
          restored: boolean(),
          requested_key: String.t(),
          key: String.t() | nil,
          path: String.t() | nil,
          size: non_neg_integer() | nil,
          compression: String.t() | nil,
          created: non_neg_integer() | nil
        }

  @doc """
  Archive the guest directory at `path` under `:key`.

  Options: `:key` (required), `:compression` (`"gzip"` by default, or `"zstd"`,
  `"estargz"`, `"none"`), `:force` to replace an existing entry.
  """
  @spec save(String.t(), String.t(), keyword()) :: {:ok, entry()} | {:error, Error.t()}
  def save(id, path, opts) do
    key = Keyword.fetch!(opts, :key)
    compression = Keyword.get(opts, :compression, "gzip")

    args = ["cache", "save", "#{id}:#{path}", "--key", key, "--json"]
    args = if compression == "gzip", do: args, else: args ++ ["--compression", compression]
    args = if Keyword.get(opts, :force, false), do: args ++ ["--force"], else: args

    with {:ok, map} <- json(args, "bsdkrun cache save") do
      {:ok, to_entry(map)}
    end
  end

  @doc """
  Restore a stored tree.

  Options: `:key` (required), `:path` (defaults to where it was saved from),
  `:restore_keys` — prefixes tried in order when the key misses.
  """
  @spec restore(String.t(), keyword()) :: {:ok, result()} | {:error, Error.t()}
  def restore(id, opts) do
    key = Keyword.fetch!(opts, :key)
    target = if path = Keyword.get(opts, :path), do: "#{id}:#{path}", else: id
    restore_keys = Keyword.get(opts, :restore_keys, [])

    args = ["cache", "restore", target, "--key", key, "--json"]
    args = if restore_keys == [], do: args, else: args ++ ["--restore-keys" | restore_keys]

    with {:ok, map} <- json(args, "bsdkrun cache restore") do
      {:ok,
       %{
         restored: Map.get(map, "restored", false),
         requested_key: Map.get(map, "requested_key", key),
         key: Map.get(map, "key"),
         path: Map.get(map, "path"),
         size: Map.get(map, "size"),
         compression: Map.get(map, "compression"),
         created: Map.get(map, "created")
       }}
    end
  end

  @doc "Every stored cache entry, newest first."
  @spec list() :: {:ok, [entry()]} | {:error, Error.t()}
  def list do
    res = Cli.run(["cache", "ls", "--json"])

    if res.exit_code == 0 do
      {:ok, res.stdout |> decode("[]") |> Enum.map(&to_entry/1)}
    else
      {:error, Error.command_failed(res.exit_code, res.stdout, res.stderr, "bsdkrun cache ls")}
    end
  end

  @doc "Remove entries by key, or every one of them with `all: true`."
  @spec remove([String.t()], keyword()) :: :ok | {:error, Error.t()}
  def remove(keys \\ [], opts \\ []) do
    args = ["cache", "rm"]
    args = if Keyword.get(opts, :all, false), do: args ++ ["--all"], else: args ++ keys

    case Cli.checked(args, "bsdkrun cache rm") do
      {:ok, _} -> :ok
      error -> error
    end
  end

  defp json(args, label) do
    res = Cli.run(args)

    if res.exit_code == 0 do
      {:ok, decode(res.stdout, "{}")}
    else
      {:error, Error.command_failed(res.exit_code, res.stdout, res.stderr, label)}
    end
  end

  defp decode(text, empty) do
    case String.trim(text) do
      "" -> Jason.decode!(empty)
      body -> Jason.decode!(body)
    end
  end

  defp to_entry(map) do
    %{
      key: Map.get(map, "key", ""),
      path: Map.get(map, "path", ""),
      compression: Map.get(map, "compression", ""),
      size: Map.get(map, "size", 0),
      created: Map.get(map, "created", 0),
      digest: Map.get(map, "digest", "")
    }
  end
end

File: sdk/clojure/src/bsdkrun/cache.clj
(ns bsdkrun.cache
  "Cached guest directories.

  Entries are keyed, so a rebuild can pick up where the last one left off:

      (require '[bsdkrun.cache :as cache])

      (let [hit (cache/restore \"web\" {:key k :restore-keys [\"deps-\"]})]
        (when-not (:restored hit)
          (sandbox/exec \"web\" [\"npm\" \"ci\"])
          (cache/save \"web\" \"/app/node_modules\" {:key k})))

  Where entries live — host disk or S3 — is host configuration, not an SDK
  concern: set `BSDKRUN_CACHE_BACKEND` / `BSDKRUN_CACHE_S3_*`, or write
  `~/.config/bsdkrun/cache.toml`."
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [bsdkrun.process :as process]))

(defn- decode
  "Parse the CLI's JSON, keywordising keys. An empty body means `empty`."
  [text empty]
  (let [body (str/trim (or text ""))]
    (json/read-str (if (str/blank? body) empty body) :key-fn keyword)))

(defn- run-json [args label empty]
  (-> (process/run! args {:label label}) :stdout (decode empty)))

(defn save
  "Archive the guest directory at `path` under `:key`.

  Options: `:key` (required), `:compression` (\"gzip\" by default, or \"zstd\",
  \"estargz\", \"none\"), `:force` to replace an existing entry. Returns the
  stored entry."
  [id path {:keys [key compression force] :or {compression "gzip"}}]
  (cond-> ["cache" "save" (str id ":" path) "--key" key "--json"]
    (not= compression "gzip") (into ["--compression" compression])
    force (conj "--force")
    :always (run-json "bsdkrun cache save" "{}")))

(defn restore
  "Restore a stored tree.

  Options: `:key` (required), `:path` (defaults to where it was saved from),
  `:restore-keys` — prefixes tried in order when the key misses. A miss is not
  an error: check `:restored` on the result."
  [id {:keys [key path restore-keys]}]
  (let [target (if path (str id ":" path) id)]
    (cond-> ["cache" "restore" target "--key" key "--json"]
      (seq restore-keys) (into (cons "--restore-keys" restore-keys))
      :always (run-json "bsdkrun cache restore" "{}"))))

(defn ls
  "Every stored cache entry, newest first."
  []
  (run-json ["cache" "ls" "--json"] "bsdkrun cache ls" "[]"))

(defn rm
  "Remove entries by key, or every one of them with `{:all true}`. Returns nil."
  ([keys] (rm keys {}))
  ([keys {:keys [all]}]
   (process/run! (if all
                   ["cache" "rm" "--all"]
                   (into ["cache" "rm"] (if (string? keys) [keys] keys)))
                 {:label "bsdkrun cache rm"})
   nil))
